Heapster tutorial (#1727)
GaloisInc/saw-script: heapster-saw/doc/tutorial/tutorial.md

# Tutorial to learn the basics of heapster-saw

This tutorial extends the current README with enough details and
examples to get anyone up to speed with using and hacking on Heapster.

<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-refresh-toc -->
**Table of Contents**

- [Tutorial to learn the basics of heapster-saw](#tutorial-to-learn-the-basics-of-heapster-saw)
    - [Building](#building)
        - [Build Saw](#build-saw)

>         - [Build the Coq backend for Saw](#build-the-coq-backend-for-saw)

Reviewer: Agreed, please use SAW and SAWScript

```
heapster_define_perm env "int64" " " "llvmptr 64" "exists x:bv 64.eq(llvmword(x))"
```

The first argument is the Heapster environment, the second is its
name, the third is its arguments (of which there are none), the fourth
is the type of value that the permission applies to, and the fifth is
its permision type. Notice how in Heapster, most of the arguments are
passed as strings. 

With this, the new permission is created and added to the
environment. Uses of this named permission are written `int64<>` where
the `<>` is the empty list of arguments, as seen in the type of `add`
above. Unfortunately, there is currently no way to print the newly defined
permissions. If you try to print the environment (`print env`) at this
point, you will only see the `llvm` definitions. We might add
functionality for showing permissions in the future.

#### 5. Type-check your program
   
Armed with the `int64` predicate, we can write the type for `add` and
ask Heapster to type check it.

```
heapster_typecheck_fun env "add" "().arg0:int64<>, arg1:int64<> -o arg0:true, arg1:true, ret:int64<>"
```

The `heapster_typecheck_fun` command takes the environment, the name

> of the function to typecheck and its permision type. The command then

Reviewer 2: ```suggestion
of the function to typecheck and its permission type. The command then
```

attempts to typecheck the function and extracts its functional
specification. The functional specification is then added to the SAW
core module `tutorial_c` with the sufix `__tuple_fun`, in this case
`add__tuple_fun`.

The function `add_mistyped`, in the same `tutorial_bc` and already
loaded in the Heapster environment, is identical to `add` so we can

> experiment with misstyping. Try running the following command

Reviewer 2: ```suggestion
experiment with mistyping. Try running the following command
```

```
heapster_typecheck_fun env "add_mistyped" "().arg0:true, arg1:int64<> -o arg0:true, arg1:true, ret:int64<>"
```

The first argument is typed as `true`, but we know it is an `i64`
which can't be proven from the trivial `true`. So this type check
should fail, but it silently terminates! What gives?

Heapster allows for the typechecker to fail in parts of
the function and the extraction will translate those parts into the
error specification. The user could then, for example, prove that
those locations are not reachable in the program, for full
correctness. Unfortunately, this means that the typechecking will fail
silently and an error won't be caught until we check the Coq
extraction, as we show in the next section.

#### 6. Writing a Coq file to prove things about the generated functional specification(s)

Once you're finished, use the following command to export all the
type-checked functions in the current environment as functional
specifications in Coq. By convention, we add a `_gen` suffix to the
filename.

```
heapster_export_coq env "tutorial_c_gen.v";
```

Open up the new `tutorial_c_gen.v` file in your examples

> directory. You should see a handfull of auto-generated imports and four

Reviewer 2: ```suggestion
directory. You should see a handful of auto-generated imports and four
```

definitions.

The first two definitions `add__tuple_fun` and `add` might have a
scary looking types, but they simplify to

```
add__tuple_fun : (bitvector 64 -> bitvector 64 -> CompM (bitvector 64)) * unit
add : bitvector 64 -> bitvector 64 -> CompM (bitvector 64)
```

That is, `add__tuple_fun` if a list of definitions, encoded as a
tuple. Saw uses these heterogeneous lists to encode functions, or
parts of them, that depend on each other.  In this case, there is only

> the function `add` function and a unit `()`, representing the end of

Reviewer 2: ```suggestion
the `add` function and a unit `()`, representing the end of
```

`errorM` in `add__tuple_fun` confirms that it was correctly
typechecked.

Notice that the converse is not true: there are some well-typed
functions that will still use `errorM` in their extracted function to,
for example, dynamically check for memory bounds. We will see those
examples in later sections. **TODO: Make sure we do this. Perhaps add
an array example?**

Before we continue, make sure to build the `.vo` file, so we can
import the definitions in `tutorial_c_gen.v` from other files. Do so with 

```
make tutorial_c_gen.vo
```

This should produce a new file `tutorial_c_gen.vo`.

##### Writting your own proofs

Open a new coq file `tutorial_c_proofs.v` in the same folder
(i.e. `heapster-saw/examples/`) and the following preamble

```
From CryptolToCoq Require Import SAWCoreScaffolding.
From CryptolToCoq Require Import SAWCorePrelude.
From CryptolToCoq Require Import CompMExtra.

(* The following line enables pretty printing*)
Import CompMExtraNotation. Open Scope fun_syntax.

Require Import Examples.tutorial_c_gen.
Import tutorial_c.
```

It first imports all the SAW functionality we need and enables pretty
printing. Then it imports our new definitions (e.g. `add`).

> Our first proof, will claim that the function `add`, produces no
> errors. Morr specifically, it says that for all inputs (that's what
> `refinesFun` quantifies) `add` allways refines to `noErrorsSpec`, that
> is it returns a pure value.

Reviewer 2: ```suggestion
Our first proof will claim that the function `add` produces no
errors. More specifically, it says that for all inputs (that's what
`refinesFun` quantifies) `add` always refines to `noErrorsSpec`. That
is, it returns a pure value.
```

```
unfold add, add__tuple_fun, noErrorsSpec.
```

Then our function `prove_refinement` will deal with the proof.

```
  prove_refinement.
Qed.
```

You can also attempt the same proof with `add_mistyped`, which
obviously will fail, since `add_mistyped` has an error.

```
Lemma no_errors_add_mistyped
  : refinesFun add_mistyped (fun _ => noErrorsSpec).
Proof.
  unfold add_mistyped, add_mistyped__tuple_fun, noErrorsSpec.
  prove_refinement. 
  (* Fails to solve the goal. *)
Abort.
```

Congratulations you have written your first Coq proofs with Heapster!

### Pointers

The next function we will type-check is a simple function that
increments the value in a pointer

```C
void incr_ptr (uint64_t *x) { *x += 1; }
```

Assuming you completed the last section, you should have interactive
saw open, the `tutorial_c.bc` loaded in the environment `env`, so
`incr_ptr` should already be in your environment, but you can double
check by printing `env`. We can then skip the steps 1-3 and go
directly to writing heapster types for the function. 

The type for this function should be

```
(). arg0:ptr((W,0) |-> int64<>) -o arg0:ptr((W,0) |-> int64<>)
```

> As before, the ghost environment is ommited and both sides of the

Reviewer 2: ```suggestion
As before, the ghost environment is omitted and both sides of the
```

implication are identical, since the function doesn't change the shape
of memory. The return value is `void`, so we can omit it or add a
trivial `ret:true`.

The permission for pointers `ptr` takes three arguments. First, it
describes the read-write modality. In this case the
pointer is writable `W`, since it will be modified. The second

> argument describes the pointer offset, here `0`. Finaly, the third

Reviewer 2: ```suggestion
argument describes the pointer offset, here `0`. Finally, the third
```

argument describes the content of the pointer, in this case a 64-bit
integer `int64<>`.

Then we can type-check the function with

```
heapster_typecheck_fun env "incr_ptr" "(). arg0:ptr((W,0) |-> int64<>) -o arg0:ptr((W,0) |-> int64<>)"
```

Finally we can generate the functional specification in Coq with 

```
heapster_export_coq env "tutorial_c_gen.v";
```

The old file should be overwritten and now contains the functional

> specification of `add`, `add_misstyped` and `incr_ptr`. As you can see

Reviewer 2: ```suggestion
specification of `add`, `add_mistyped` and `incr_ptr`. As you can see
```

the definition of `incr_ptr__tuple_fun` has no references to `errorM`,
so we know it was correctly type checked.

You will have to generate the `.vo` again to write proofs about
`incr_ptr`. After you do so, we can easily prove that `incr_ptr`
produces no errors.

```
Lemma no_errors_incr_ptr
  : refinesFun incr_ptr (fun _ => noErrorsSpec).
Proof.
  unfold incr_ptr, incr_ptr__tuple_fun.
  prove_refinement.
Qed.
```

### Structs

The next function we will type-check deals with structs. In our
example, we defined a function that can compute the norm of a 3D
vector

``` C

> // Struct that represents the three coordinates fora 3D vector

Reviewer 2: ```suggestion
// Struct that represents the three coordinates for a 3D vector
```

// || (x,y,z) || = x^2+y^2+z^2
uint64_t norm_vector (vector3d *v) { return (v->x * v->x + v->y * v->y + v->z * v->z); }
```

Again, we assume that you still have the definitions from the previous
sections so we can start defining the type for the function.

Let's start by defining a predicate for `vector3d` like so

```
heapster_define_perm env "vec3d" "rw:rwmodality" "llvmptr 64" "ptr((rw,0) |-> int64<>) * ptr((rw,8) |-> int64<>) * ptr((rw,16) |-> int64<>)"
```

First, notice that we added an arguments `rw` of type
`rwmodality`. This is such that we can control if the vector is
readable or writable. Second, notice that the predicate still applies
to 64-bit pointers. Finally, the type describes three integers, all
with the read-write modality given by the argument `rw`, at offsets
`0`, `8` and `16` and each with an `int64`.

Then we can define the type of `norm_vector` as

```
(). arg0:vec3d<R> -o arg0:vec3d<R>, ret:int64<>
```

which says that the function takes a readable 3D vector, and returns
an integer. Notice that the `arg0` on the right hand side could also
be written as `arg0:true`. However, we still want to express that the

> function does change that memory so we make it explicit.

Reviewer 2: Wait, why does the function change the memory? I wouldn't have expected that since it's using the `R` modality.

Reviewer: Actually, yes, `norm_vector` should just have `arg0:true` as its output permission. The convention is that any copyable permissions in the input, such as `eq` or read-only pointer permissions, do not need to get returned in the output.

print "Type checked add."

heapster_typecheck_fun env "add_mistyped" "().arg0:true, arg1:int64<> -o arg0:true, arg1:true, ret:int64<>"
print "Type checked add_mistyped. This will produce an error in the output."

heapster_typecheck_fun env "incr_ptr" "(). arg0:ptr((W,0) |-> int64<>) -o arg0:ptr((W,0) |-> int64<>)"
print "Type checked incr_ptr."

heapster_define_perm env "vec3d" "rw:rwmodality" "llvmptr 64" "ptr((rw,0) |-> int64<>) * ptr((rw,8) |-> int64<>) * ptr((rw,16) |-> int64<>)"
heapster_typecheck_fun env "norm_vector" "(). arg0:vec3d<R> -o arg0:vec3d<R>, ret:int64<>"
print "Type checked incr_ptr."

heapster_export_coq env "tutorial_c_gen.v";
print "Export to Coq."

print "Done."
```

then you can process it with just

```
% cabal run saw -- tutorial_c.saw
Up to date

[16:41:49.222] Loading file "/Users/Santiago/Projects/saw-script/heapster-saw/examples/tutorial_c.saw"
[16:41:49.230] File loaded
[16:41:49.245] Type checked add.
[16:41:49.249] Type checked add_mistyped. This will produce an error in the output.
[16:41:49.257] Type checked incr_ptr.
[16:41:49.312] Type checked norm_vector.
[16:41:49.329] Export to Coq.
[16:41:49.329] Done.
```

### Arrays

> We will briefly explore arrays, which have slightly more intersting

Reviewer 2: ```suggestion
We will briefly explore arrays, which have slightly more interesting
```

memory restrictions. Namely, array access must be in bounds. The code
in this section are already generated for you and you can find them,
together with more examples in the files `arrays.c`, `arrays.saw`,
`arrays_gen.v` and `arrays_proofs.v`.

> We will consider the function `zero_array` which zero's out all the

Reviewer 2: ```suggestion
We will consider the function `zero_array` which zeroes out all the
```

values in an array (see code in `arrays.c`).

```C
void zero_array (int64_t *arr, uint64_t len) {
  for (uint64_t i = 0; i < len; ++i) {
    arr[i] = 0;
  }
}
```

The type for this function is relatively simple, it only assumes that
`len` is actually the length for the given array `arr`. This is
achieved by used a shared or ghost variable `len` which is both the

> lenght of the array and equal to the second argument (see the code in `arrays.saw`)

Reviewer 2: ```suggestion
length of the array and equal to the second argument (see the code in `arrays.saw`)
```

```
heapster_typecheck_fun env "zero_array"
  "(len:bv 64). arg0:int64array<len>, arg1:eq(llvmword(len)) -o \
              \ arg0:int64array<len>, arg1:true, ret:true";
```

Heapster also expects a loop invariant hint for every loop. Loop
invariants look just like function types, taking the loop variables as
arguments. In this case the loop introduces a new variable `i` which

> is the ofset into the array. We represent that with a new ghost

Reviewer 2: ```suggestion
is the offset into the array. We represent that with a new ghost
```

variable

```
heapster_typecheck_fun env "zero_array_from"
  "(len:bv 64, off:bv 64). arg0:int64array<len>, arg1:eq(llvmword(len)), arg2:eq(llvmword(off)) -o \
                         \ arg0:int64array<len>, arg1:true, arg2:true, ret:true";
```

Certainly function correctness must ensure that all the writes to the
array (i.e. `arr[i] = 0`) happen within bounds. However this is a
dynamic property which is not part of type-checking. Instead, Heapster

> adds dynamic checks on the extracted code which we will see in the coq

Reviewer 2: ```suggestion
adds dynamic checks on the extracted code which we will see in the Coq
```

code. 

> Let's go to `arrays_gen.v` (which has already been generatred for you)

Reviewer 2: ```suggestion
Let's go to `arrays_gen.v` (which has already been generated for you)
```

and look for the definition of `zero_array__tuple_fun`. You will
notice that it calls `errorM` twice but, in this case that's not a
sign of a typing error! Heapster includes these errors to catch

> out-of-bounds array accesses and a unrepresentable indices (i.e. index

Reviewer 2: ```suggestion
out-of-bounds array accesses and unrepresentable indices (i.e. index
```

```
zero_array__tuple_fun = 
    (* ... Some ommited code ... *) 
    LetRec f:=
     (fun (e1 e2 : int64) (p1 : Vector int64 e1)
        (_ _ _ : unit : Type) =>
        (* ... Some ommited code ... *) 
        let var__4 := (0)[1] in
        let var__6 := e2 < e1 in
		(*... Some ommited code ...*)
	  if
       not
         ((if var__6 then 1 else var__4) ==
          var__4)
      then
       if
        ((17293822569102704640)[64] <= e2) &&
        (e2 <= (1152921504606846975)[64])
       then
        If e2 <P e1
        Then f e1 (e2 + (1)[64])
               (p1 [e2 <- (0)[64]]) tt tt tt
        Else errorM "ghost_bv <u top_bv"
       else
        errorM "Failed Assert at arrays.c:61:5"
      else returnM p1, tt)
      InBody (f e0 (0)[64] p0 tt tt tt), tt))
```
 
The arguments `e1`, `e2` and `p1` correspond to the length `len`, the
offset `i`, and the array `arr`. Notice most of that the function
performs several tests before executing any computation. The first two
tests, which are within `if then else` blocks, check that the offset
is less than the array length `e2 < e1`, and that the index `i` is
within the bounds of machine integers. The former is part of the loop

> condition, the second is added by heapster to ensure the ghost

Reviewer 2: ```suggestion
condition, the second is added by Heapster to ensure the ghost
```

variable represents a machine integer.

The last check, which is within uppercase `If Then Else` (which
handles option types) is not part of the original function but
inserted by Heapster. It checks that the array access is within bounds
`e2 <P e1`. The operator `<P` returns an optional proof of the array
access being safe. If the check fails, the function reports a runtime
error `errorM "ghost_bv <u top_bv"`.

> If all the checks pass, then the function simply recursive calls
> itself, with the next index and array with a new entry zeroed out.

Reviewer 2: ```suggestion
If all the checks pass, then the function simply calls
itself recursively, with the next index and array with a new entry zeroed out.
```

`Import CompMExtraNotation. Open Scope fun_syntax.` after the files
imports. Then, after processing the file you can write `Print
zero_array__tuple_fun`, after the function has been defined, and you
should see the pretty printed version appear in the response. You
don't really modify `arrays_gen.v` so you can do this exploration in
`arrays_proofs.v` or just remember to clean up `arrays_gen.v` after
you are done.

We shall now prove that this function in fact should never return an
error if the inputs are correct. Go to `array_proofs.v` and look at the proof 

```
Lemma no_errors_zero_array
  : refinesFun zero_array (fun x _ => assumingM (zero_array_precond x) noErrorsSpec).
Proof
```

It claims that, assuming the precondition `zero_array_precond` is
satisfied, then the function `zero_array` produces no errors. The
precondition simply says that the length of the array is within
computable integers.

```
Definition zero_array_precond x
  := 64 <= x /\ x <= bvMem_hi.
```

We will not go into detail about the proof, but notice that the
important steps are handled by custom automation. Namely we use the
commands `prove_refinement_match_letRecM_l` and `prove_refinement` to

> handle refinments with and without recursion (respectively). The rest

Reviewer 2: ```suggestion
handle refinements with and without recursion (respectively). The rest
```

  %struct.list64_t* @find_elem(i64 %0, %struct.list64_t* %1)
  %struct.list64_t* @sorted_insert(i64 %0, %struct.list64_t* %1)
  %struct.list64_t* @sorted_insert_no_malloc(%struct.list64_t* %0,
                                             %struct.list64_t* %1)
```

#### 4. Writing heapster types for your functions

We can check in the environment `env` for the LLVM type of the function we
are type checking. 

```LLVM
i64 @is_elem(i64 %0, %struct.list64_t* %1)
```

The function `is_elem` takes a 64-bit integer and a list of 64-bit
integers, and returns another 64-bit integer. After the return, we
don't care about the inputs, so we can write the type like this

```
().arg0:int64<>, arg1:List64<R> -o arg0:true, arg1:true, ret:int64<>
```

But we will need to define the predicates for `int64` and `List64`

##### Defining list permissions

We know how to define `int64`, as we did before, 

```
sawscript> heapster_define_perm env "int64" " " "llvmptr 64" "exists x:bv 64.eq(llvmword(x))"
```

but we need a new predicate for lists.  Before we look at the
definition of a `List64<rw>` lets focus on its permission type. First
of all, `List64<rw>` takes a single argument `rw:rwmodality` which

> determines if the list is readable or writable, jus like 3D

Reviewer 2: ```suggestion
determines if the list is readable or writable, just like 3D
```

vectors. It's type should look something like this

```
["eq(llvmword(0))", "ptr((rw,0) |-> int64<>) * ptr((rw,8) |-> List64<rw>)"]
```

> the definition shows the diferent cases for a list, separated by a
> coma. In the first case, a `List64` can be a null pointer, expressed
> with the type `eq(llvmword(0))`. In the second case, a list is a
> pointer where offset `0` is the head, an `Int64`, and offset `8` it's
> the tail, is recursively a `List64<rw>`. In the later case,
> both elements are tagged with `rw`, describing if they are readable or
> writable, as determined by the argument to `List64<rw>`.

Reviewer 2: ```suggestion
The definition shows the diferent cases for a list, separated by a
comma. In the first case, a `List64` can be a null pointer, expressed
with the type `eq(llvmword(0))`. In the second case, a list is a
pointer where offset `0` is the head, an `Int64`, and offset `8` is
the tail, is recursively a `List64<rw>`. In the later case,
both elements are tagged with `rw`, describing if they are readable or
writable, as determined by the argument to `List64<rw>`.
```

To define [permissions](doc/Permissions.md) which can describe
unbounded data structures, you can use the

> `heapster_define_recursive_perm` command. here is how to use the

Reviewer 2: ```suggestion
`heapster_define_recursive_perm` command. Here is how to use the
```

command to define lists.

```
heapster_define_recursive_perm
  env
  "List64"
  "rw:rwmodality"
  "llvmptr 64"
  ["eq(llvmword(0))", "ptr((rw,0) |-> int64<>) * ptr((rw,8) |-> List64<rw>)"]
  "List_def" "foldList" "unfoldList";
```

Its first four arguments are the same as for `heapster_define_perm`,
namely the environment, the name, the arguments and the type of value

> that the permission applies to. The fifth argument is its permision

Reviewer 2: ```suggestion
that the permission applies to. The fifth argument is its permission
```

    - [Looking under the hood](#looking-under-the-hood)
        - [Heapster commands and environments](#heapster-commands-and-environments)
        - [Permissions](#permissions)
        - [Type-checking](#type-checking)

<!-- markdown-toc end -->

## Building

We'll start by building everything you need to use Heapster. All the
commands here are with respect to the top-level saw-script directory.

### Build Saw 

You will need to follow the instructions in the top-level README to
download or build a SAW binary, of which Heapster is a part. In
particular, make sure you follow the instructions to install Z3. Once `./build.sh`
succeeds it should report 

```bash
COPIED EXECUTABLES TO /Path/To/Saw/saw-script/bin.
```

If everything is installed correctly you should be able to run saw

```bash
cabal run saw
```

This should open the saw interactive session. It should show you a
pretty SAW logo with the version number. We will learn more about the
interactive session later. For now you can quit the session with `:quit` or
`:q` like so:

```bash
sawscript> :quit
```

### Build the Coq backend for Saw

In this tutorial we will also interact with Heapster's Coq output. So you'll need

> to follow the instructions in the README in the `saw-core-coq` subdirectory.

Reviewer: Please add a link to that README

Author: What do you have in mind for this link? Do you mean a hyperlink to the README in the master branch in github?

The location would be different if the tutorial is being read online or locally. That's why I tend to use local commands when I can.

Specifically, after installing the dependencies, you will need to run the
following (from this directory):
```bash
cd /saw-core-coq/coq
make
```

It is expected that you will see a large number of warnings, but the
build should complete without any errors.

**TODO: How do we check if this is properly installed before continuing?**

> Before continuing, return to the top-level directory with `cd ..`.

Reviewer: This should be `cd ../..`

The process will generally involve

- [1. Generating LLVM bitcode](#1-generating-llvm-bitcode)
- [2. Run the SAW interpreter with Heapster](#2-run-the-saw-interpreter-with-heapster)
- [3. Load the file and extract the two function specifications.](#3-load-the-file-and-extract-the-two-function-specifications-1)
- [4. Writing heapster specifications for your functions](#4-writing-heapster-specifications-for-your-functions)
- [5. Writing a SAW script to type-check your code with respect to the sepcification](#5-writing-a-saw-script-to-type-check-your-code-with-respect-to-the-sepcification)
- [6. Writing a Coq file to prove things about the generated functional specification(s)](#6-writing-a-coq-file-to-prove-things-about-the-generated-functional-specifications)

Just like with SAW, Heapster can be processed in batch. To do so, you
can combine steps 2-6 in a `.saw` file and use SAW's batch processing.

### Running an example

This section will walk through the process of using Heapster to write
and verify some C code. Specifially, we want to verify the function
`is_elem`, which tests if a specific value is in a list. The function,

> together with others, can be found in `examples/linked_list.c`.

Reviewer: Instead of expecting the user to be in the `heapster-saw` directory for all this tutorial, you should have them be in `heapster-saw/examples`, so that the user can access the Makefile listed there. Maybe put an explicit `cd heapster-saw/examples` command in section 1 (or 0)?

Author: If it makes sense, I'll add it at the beginning of the Heapster section, since installation and running Saw makes more sense from the top level directory.

```C
typedef struct list64_t {
  int64_t data;
  struct list64_t *next;
} list64_t;

int64_t is_elem (int64_t x, list64_t *l) {
  if (l == NULL) {
    return 0;
  } else if (l->data == x) {
    return 1;
  } else {
    return is_elem (x, l->next);
  }
}
```

#### 1. Generating LLVM bitcode

Just like with SAW, we want to work with the LLVM bitcode (`.bc`), so
you can generate it just like before. Note that we have already
included the binary for all the examples so you don't really need to
run this command.

```bash
clang -g -c -emit-llvm -o examples/linked_list.bc examples/linked_list.c
```

(The alternative command `make examples/linked_list.bc` won't work

> because the shorcut is not defined in the Makefile of heapster.)

Reviewer: I'm confused by this comment, because `make linked_list.bc` works for me (when run in the `heapster-saw/examples` directory)

Be aware that the resulting bitcode may depend on your `clang` version and your
operating system. In turn, this means the Heapster commands in your SAW script
and the proofs in your Coq file may also be dependent on how and where the
bitcode is generated. If you find an incompatibility, please report it.

#### 2. Run the SAW interpreter with Heapster

We start by running saw with `cabal run saw`. Once SAW is loaded, you
can load all the Heapster commands with

```bash
sawscript> enable_experimental
```

> If you print the environment now (wiht `:env`) you will notice a new

Reviewer: Typo: "wiht" -> "with"

heapster_assume_fun_rename_prim      heapster_gen_block_perms_hint
heapster_block_entry_hint            heapster_get_cfg
heapster_define_irt_recursive_perm   heapster_init_env
heapster_define_irt_recursive_shape  heapster_init_env_debug
heapster_define_llvmshape            heapster_init_env_for_files
heapster_define_opaque_llvmshape     heapster_init_env_for_files_debug
heapster_define_opaque_perm          heapster_init_env_from_file
heapster_define_perm                 heapster_init_env_from_file_debug
heapster_define_reachability_perm    heapster_join_point_hint
heapster_define_recursive_perm       heapster_parse_test
heapster_define_rust_type            heapster_print_fun_trans
heapster_define_rust_type_qual       heapster_set_debug_level
heapster_export_coq                  heapster_set_translation_checks
heapster_find_symbol                 heapster_typecheck_fun
heapster_find_symbol_commands        heapster_typecheck_fun_rename
heapster_find_symbol_with_type       heapster_typecheck_mut_funs
```

You can then use `:?` to see further information for each of them.

#### 3. Load the file and extract the two function specifications.

> To load a file into heapster you can use `heapster_init_env`. Let's

Reviewer: You should maybe introduce `heapster_init_env_from_file` first, because almost all of the examples (including specifically the `linked_list` example) use it, and then you can add an extra few sentences mentioning that `heapster_init_env` is a special case

> check it's documentation first

Reviewer: Typo: "it's" -> "its"

that can be printed wiht `:env`).

Let's load our module with

```
env <- heapster_init_env "linked_list" "examples/linked_list.bc";
```

we have created a new Heapster environment that we can explore. 

```
sawscript> print env
[20:19:48.436] Module: examples/linked_list.bc
Types:
  %struct.list64_t = type { i64, %struct.list64_t* }

Globals:

External references:
  declare default void @llvm.dbg.declare(metadata, metadata,
                                         metadata)
  declare default i8* @malloc(i64)

Definitions:
  i64 @is_head(i64 %0, %struct.list64_t* %1)
  i64 @is_elem(i64 %0, %struct.list64_t* %1)
  i64 @any(i64(i64)* %0, %struct.list64_t* %1)
  %struct.list64_t* @find_elem(i64 %0, %struct.list64_t* %1)
  %struct.list64_t* @sorted_insert(i64 %0, %struct.list64_t* %1)
  %struct.list64_t* @sorted_insert_no_malloc(%struct.list64_t* %0,
                                             %struct.list64_t* %1)
```

The heapster environment contains all the types, global definitions,
external references and functions from the loaded module.

#### 4. Writing heapster specifications for your functions

##### Pen and paper specification

> 

Reviewer: You should use the term "Heapster types" instead of "Heapster specifications", because, like you say later, Heapster types aren't full functional specifications, but instead mostly deal with memory shapes and whether LLVM values are pointers or word values.

Before we go foward using Heapster, let's think about what the
specification for `is_elem` should be. As we can check in the
environment `env`, `is_elem` takes a 64-bit integer the "element".

```LLVM
i64 @is_elem(i64 %0, %struct.list64_t* %1)
```

> Hence, precondition to the function should reflect these two inputs

Reviewer: It's a little confusing that you introduce some notation here that is close to but not quite the same as Heapster notation. Maybe, instead of trying to first define the type in an abstract way and refining it in the next sub-section to the Heapster syntax, you could just give the Heapster type for `is_elem` and explain each of its pieces

Author: Yes this makes sense. Since we are also adding a simpler example, I think I can directly show the type and explain it.

true, ret:int64`, where `true` is a predicate the is always satisfied.
 
Then, using the linear implication `-o` we can express the
specification of `is_elem` as

```
arg0:int64, arg1: list int64 -o arg0:true, arg1:true, ret:int64
```

Notice that this is a memory-safety specification, not a correctness
predicate. We can reason about the correctness of the function once we
extract it as a functional program in Coq.

In the next sections, we shall define the predicates necessary to
write this specification in Heapster.

##### Defining permission predicates

One of the simplest Heapster commands is `heapster_define_perm`, which defines
a new named permission which can then be used in Heapster types. As an
example, the following defines the permission which describes a 64-bit integer
value:

```
heapster_define_perm env "int64" " " "llvmptr 64" "exists x:bv 64.eq(llvmword(x))"
```

The first argument is the Heapster environment, the second is its
name, the third is its arguments (of which there are none), the fourth
is the type of value that the permission applies to, and the fifth is
its definition. the new permission is created and added to the

> environment.

Reviewer: Mention that uses of named permissions are written like `int64<>`, where the `<>` is the empty list of arguments.

Unfortunately, there is currently no way to print the newly defined
permissions. If you try to print the environment (`print env`) at this
point, you will only see the `llvm` definitions. We might add
functionality for showing permissions in the future.

The permission predicate for lists is a little more complicated
because it requires a recursive definition. To define
[permissions](doc/Permissions.md) which can describe unbounded data
structures, you can use the `heapster_define_recursive_perm`
command. As an example, here is how to describe a linked list of
64-bit words using this command:

```
heapster_define_recursive_perm
  env
  "List64"
  "rw:rwmodality"
  "llvmptr 64"
  ["eq(llvmword(0))", "ptr((rw,0) |-> int64<>) * ptr((rw,8) |-> List64<rw>)"]
  "List (Vec 64 Bool)"
  "foldList (Vec 64 Bool)"
  "unfoldList (Vec 64 Bool)";
```

> Its first four arguments are the same as for `heapster_define_perm`,

Reviewer: Explain the `rw` argument, because this is the first example that uses it

its fifth argument contains its different inductive cases (in this

> case, a `List64` is either a null pointer, or a pointer to an `Int64`

Reviewer: Expand the part about the cases into its own 1-2 sentences describing the two different cases and what their syntax means, even if briefly. For instance, say something like "In this case, a `List64` is either a null pointer, which is expressed with the type `eq(llvmword(0))` that declares that it equal to the word value 0, or it is a pointer where offset 0 points to a value satisfying the `int64<>` type and where offset 8 points to a value satisfying the `List64<rw>` permission, i.e., it is recursively a `List64<rw>`.

Reviewer: Also reiterate that the syntax `List64<rw>` is the result of applying the `List64` named permission to the single argument `rw`.

and another `List64`), and its final three arguments are its
translation into SAW core. Here the SAW core definitions used are from
the SAW core prelude which are included by default when Heapster
autogenerates a a module for you. If you need new SAW core
definitions, you will need to use the following command instead of
`heapster_init_env`:

```
env <- heapster_init_env_from_file "my_file.sawcore" "my_file.bc";
```

See this [additional documentation](doc/Permissions.md) for a
reference on the syntax and meaning of heapster permissions.

#### 5. Writing a SAW script to type-check your code with respect to the sepcification
   
We begin by converting the pen-and-paper specification of `is_elem`
into a Heapster specification. First, every heapster type begins with
the declaration of a ghost environment. Our specification doesn't use

> any ghost variable so that is `().` Moreover, for every predicate we

Reviewer: This explanation of every predicate requiring a read-only or read-write modality is not quite right. The reason we write `List64<R>` here is because `List64` was defined to take a single argument of type `rwmodality`. Read-write access to a list would require the type `List64<W>`. You can't use `<>` in place of `<R>` because the former is 0 arguments and the latter is 1.
